src/binary/decode.rs

```diff
@@ -0,0 +1,2179 @@
+use crate::ast::*;
+use id_arena::Id;
+use std::io::Read;
+
+/// A trait implemented by every Web IDL bindings thing that can be decoded from
+/// an input stream.
+pub(crate) trait Decode {
+    /// The output of decoding `Self` from the input stream.
+    ///
+    /// Since the custom section's arenas and AST nodes typically live inside
+    /// the `cx`, this is usually some sort of typed ID pointing into an arena.
+    type Output;
+
+    /// Decode an instance of `Self` into the given `cx`.
+    fn decode<Cx, R>(cx: &mut Cx, r: &mut R) -> Result<Self::Output, failure::Error>
```

Comment:
This signature is really throwing me for a spin relative to what I'd expect from a `Decode` I think. I sort of just assumed that this would return `failure::Result<Self>` and then got pretty confused at all the usage of `Ignore` and `Output = ()`. I found it somewhat difficult to see what's being inserted where. 

Would it be possible to remove the `Output` associated type and change this to just return `Self`? I think that might help making it a bit more clear for me and having one location to insert types, for example, vs multiple locations that insert types.

Reply:
Unfortunately, `Self` as the return type doesn't work so well with the arena approach. I did remove the `Cx` type param and am looking into removing the `Read` param too, which should help clean things up.

Reply:
I investigated this a bit but the hard part comes from thigns that reference other things by index (like webidl types) where we haven't yet finished constructing a whole `WebidlTypes` and its arena. This is why we sometimes decode by mutating the under-construction custom section in the cx instead of always returning the decoded value. I don't see a nice/easy way to work around this.